### DISCUSSION TITLE
Conditionally hide / disable template or fields

### DISCUSSION THREAD

**Discussion by cjmcintyre** [tinacms/tinacms#2831] [Ideas]
With the latest changes I've been trying to implement a conditionally hide / disable template or fields.
**The problem**

1) We have a hero component on every page.
![image](https://user-images.githubusercontent.com/18439866/164395881-90d796f3-47bf-4109-8a4d-9d4e9ba45d97.png)

2) We have multiple "versions" for a hero component and each of those have similar properties but not all the same and I'd like to hide or show certain properties depending on what is chosen in "Version". At the moment I just have multiple components built in and just a switch case and just don't use that data if I don't require it, but this obviously leaves my user being able to "Fill in" the field and probably confused where it is being displayed.
![image](https://user-images.githubusercontent.com/18439866/164396487-7abc131a-7230-412c-8a7c-94b37e3e770c.png)

**My current solution?**

After a discussion was raised on Discord I asked whether It's possible to do it in there and **llibdude** offered a hint towards using the component: **({ form, field, input })**

Here is what I've come up with.. and it _kind of works._

```javascript
function getComponent(form) {
  const version = (form.field.label).toLowerCase(); // Herov1, Herov2, Herov3
  const array = form.tinaForm.values.blocks;
  const chosenHeroVersion = array.find((o) => o.heroVersion);

  if (chosenHeroVersion?.heroVersion === version) {
    form.field.component = "group";
    return form.field.component;
  } else {
    return <div></div>;
  }
}

const testBlockSchema: TinaTemplate = {
  name: "Test",
  label: "Test",
  fields: [
    {
      type: "string",
      label: "Hero Version",
      name: "heroVersion",
      options: [
        { label: "Herov1", value: "herov1" },
        { label: "Herov2", value: "herov2" },
        { label: "Herov3", value: "herov3" },
      ],
      ui: {
        defaultValue: "herov3",
      }
    },
    {
      type: "object",
      name: "herov1",
      label: "Herov1",
      ui: {
        component: (form) => getComponent(form),
      },
      fields: [
        {
          type: "string",
          label: "Title",
          name: "title",
        },
      ]
    },
    {
      type: "object",
      name: "herov2",
      label: "Herov2",
      ui: {
        component: (form) => getComponent(form),
      },
      fields: [
        {
          type: "string",
          label: "Title",
          name: "title",
        },
      ]
    },
  ],
};

```

1) **herov3 Is blank, this is expected**
![image](https://user-images.githubusercontent.com/18439866/164402714-381a5b5c-b58f-4328-b587-62c347d55c88.png)

2) herov2 selected, it appears
![image](https://user-images.githubusercontent.com/18439866/164402887-2cef9077-5ba9-47c6-ac14-717dccf2bd1f.png)

3) herov1 selected.. it appears, however herov2 also stays..
![image](https://user-images.githubusercontent.com/18439866/164403009-70bb7c92-30b7-40ff-b81f-84bc81e79cb8.png)

If anyone feels like they can offer some assistance I would be greatly appreciative.

**Comment by jeffsee55** [comment]
@cjmcintyre why don't you use templates for this?

```
const testBlockSchema: TinaTemplate = {
  name: 'Test',
  label: 'Test',
  fields: [
    {
      name: 'hero',
      label: 'Hero',
      type: 'object',
      templates: [
        {
          name: 'herov1',
          label: 'Herov1',
          fields: [
            {
              type: 'string',
              label: 'Title',
              name: 'title',
            },
          ],
        },
        {
          name: 'herov2',
          label: 'Herov2',
          fields: [
            {
              type: 'string',
              label: 'Some Title',
              name: 'someTitle',
            },
          ],
        },
      ],
    },
  ],
}
```

**Reply by cjmcintyre** [reply]
Hi @jeffsee55 
Implementing the code above throws `Error: blocks without list true is not yet supported`. Adding list: true would allow a user to do the following..

![image](https://user-images.githubusercontent.com/18439866/164818495-9584a95c-fd10-4f38-8875-e70719c9e2db.png)

I guess I could then _validate_ to check if it is > 1..

**Comment by jaschaio** [comment]
There was an issue for this but it [unfortunately got closed by the stale bot](https://github.com/tinacms/tinacms/issues/1678). I think this feature would be incredible useful for more complex use cases.

**Comment by Jason-Jay-Mason** [comment]
Hey there! I thought I would add my solution here for the community. This is how I achieved conditional field in 1.4.40:

First you need this conditional field factory function:

```typescript
import React from "react";

export function ConditionalField(
 fieldName: string,  // the name of the triggering field
 showValue: string,  // what value triggers 'isVisible'
 Component: React.FC<any>  // the tina component to render
): React.FC<any> {
 return (props) => {
   const isVisible: boolean = React.useMemo(() => {
     const val = props?.tinaForm?.finalForm?.getState().values[fieldName];
     return val && `${val}` === showValue;
   }, [props]);

   return (
     <div className={isVisible ? "pb-5" : "hidden"}>
       <Component {...props} />
     </div>
   );
 };
}
```
And the example usage:
```typescript
export const example: Collection = {
  name: "example",
  label: "Example",
  path: "content/example",
  format: "md",
  fields: [
    {
      name: "triggeringField",
      label: "Triggering Field",
      type: "string",
      ui: {
        component: "select",
        //@ts-ignore
        options: ["No", "Yes"],
      },
    },
    {
      name: "conditionalField",
      label: "Conditional Field",
      type: "string",
      ui: {
        component: ConditionalField("triggeringField", "Yes", SelectField),
      },
    },
  ],
};
```
